README
A.L.F.R.E.D - Almighty Lightweight Fact Remote Exchange Daemon
--------------------------------------------------------------

"alfred is a user space daemon to efficiently[tm] flood the network with
 useless data - like vis, weather data, network notes, etc"
   - Marek Lindner, 2012

Introduction
------------

alfred is a user space daemon for distributing arbitrary local information over
the mesh/network in a decentralized fashion. This data can be anything which
appears to be useful - originally designed to replace the batman-adv
visualization (vis), you may distribute hostnames, phone books, administration
information, DNS information, the local weather forecast ...

alfred runs as daemon in the background of the system. A user may insert
information by using the alfred binary on the command line, or use special
programs to communicate with alfred (done via unix sockets). alfred then takes
care of distributing the local information to other alfred servers on other
nodes. This is done via IPv6 link-local multicast, and does not require any
configuration. A user can request data from alfred, and will receive the
information available from all alfred servers in the network.

Compilation
-----------

alfred depends on:
 * librt (usually part of libc)
 * IPv6 support in the kernel/host system
 * jansson (C library for working with JSON data)

To compile alfred, simply type:

 $ make

This will compile both alfred and vis. To install, use

 $ make install

(with the right privileges).

If you don't want to compile vis, add the directive CONFIG_ALFRED_VIS=n:

 $ make CONFIG_ALFRED_VIS=n 
 $ make CONFIG_ALFRED_VIS=n install

Usage
-----

First, alfred must run as daemon (server) in background to be used. This can
either be done by some init-scripts from your distribution (if you have
received alfred as a package with your distribution). Please see their
documentation how to configure alfred in this case. In any event, you can
still run alfred from the command line. The relevant options are (for a full
list of options, run alfred -h):

  -i, --interface             specify the interface to listen on
  -b                          specify the batman-adv interface configured on
                              the system (default: bat0). use 'none' to disable
                              the batman-adv based best server selection
  -m, --master                start up the daemon in master mode, which
                              accepts data from slaves and synces it with
                              other masters

The -b option is optional, and only needed if you run alfred on a batman-adv
interface not called bat0, or if you don't use batman-adv at all
(use '-b none'). In this case, alfred will still work but will not be able to
find the best next master server based on metrics.

alfred servers may either run as master or slave in the network. Masters will
announce their status via broadcast, so that slaves can find them. Slaves will
then send their data to their nearest master (based on TQ). Masters will
exchange their data (which they have received from slaves or got on their own)
with other masters. By using masters and slaves, overhead can be reduced while
still keeping redundancy (by having multiple masters). Obviously, at least one
master must be present in the network to let any data exchange happen. Also
having all nodes in master mode is possible (for maximum decentrality and
overhead).

To put it together, let us start alfred in master mode on our bridge br0
(assuming that this bridge includes the batman interface bat0):

 $ alfred -i br0 -m

Now that the server is running, let us input some data. This can be done by
using the alfred binary in client mode from the command line:

 $ cat /etc/hostname | alfred -s 64

This will set the hostname as data for datatype 64. Note that 0 - 63 are
reserved (please send us an e-mail if you want to register a datatype), and can
not be used on the commandline. We skipped the version parameter allowing you
to assign a version to your data which can be filtered by other alfred users.
Skipping the parameter entirely has the same effect as setting the parameter
to 0 ('-V 0').

After the hostname has been set on a few alfred hosts, the can be retrieved again:

 $ alfred -r 64
{
  "fe:f1:00:00:01:01": {
    "hostname": "alfred",
    "ip_address": [192, 168, 1, 23]
  },
  "fe:f1:00:00:02:01": "OpenWRT-node\n",
  "fe:f1:00:00:03:01": [
    0,
    1,
    255
  ]
}

Values will be expressed as JSON objects, UTF-8 strings or byte arrays. First,
alfred tries to interprete the data as a JSON object. If this succeeds, it'll
set the value to the JSON object. If this fails, it'll try expressing the data
as an UTF-8 string. As a last resort, the data will be expressed as an byte
array.

Note that the information must be periodically written again to alfred, otherwise
it will timeout and alfred will forget about it (after 10 minutes).

One final remark on terminology: If we talk about "servers" and "clients" in
alfred, we mean the local processes on one machine which talk to each other via
unix sockets (client connects and talks to servers). On the other hand, "slaves"
and "masters" are the roles alfred can take over in the network between different
machines (slaves send information to masters).

Vis
---

Vis can be used to visualize your batman-adv mesh network. It read the neighbor
information and local client table and distributes this information via alfred
in the network. By gathering this local information, any vis node can get the
whole picture of the network.

Vis, similar to to alfred, combines server (daemon) and client functionality in
the 'vis' binary. The vis server must be started to let vis work:

 $ vis -i bat0 -s

This server will read the neighbor and client information from batman-adv every
10 seconds and set it in alfred via unix socket. Obviously, the alfred server
must run too to get this information set.

To get a graphviz-compatible vis output, simply type:
 $ vis
digraph {
        subgraph "cluster_fe:f0:00:00:04:01" {
                "fe:f0:00:00:04:01"
        }
        "fe:f0:00:00:04:01" -> "fe:f0:00:00:05:01" [label="1.000"]
        "fe:f0:00:00:04:01" -> "fe:f0:00:00:03:01" [label="1.004"]
        "fe:f0:00:00:04:01" -> "00:00:43:05:00:04" [label="TT"]
        "fe:f0:00:00:04:01" -> "fe:f1:00:00:04:01" [label="TT"]
        subgraph "cluster_fe:f0:00:00:02:01" {
                "fe:f0:00:00:02:01"
        }
        "fe:f0:00:00:02:01" -> "fe:f0:00:00:03:01" [label="1.000"]
        "fe:f0:00:00:02:01" -> "fe:f0:00:00:01:01" [label="1.008"]
        "fe:f0:00:00:02:01" -> "fe:f0:00:00:08:01" [label="1.000"]
        "fe:f0:00:00:02:01" -> "fe:f1:00:00:02:01" [label="TT"]
        "fe:f0:00:00:02:01" -> "00:00:43:05:00:02" [label="TT"]
        subgraph "cluster_fe:f0:00:00:08:01" {
                "fe:f0:00:00:08:01"
        }
[...]
}

For a json formatted output, use:

 $ vis -f json
{ "primary" : "fe:f0:00:00:04:01" }
{ "router" : "fe:f0:00:00:04:01", "neighbor" : "fe:f0:00:00:05:01", "label" : "1.000" }
{ "router" : "fe:f0:00:00:04:01", "neighbor" : "fe:f0:00:00:03:01", "label" : "1.008" }
{ "router" : "fe:f0:00:00:04:01", "gateway" : "00:00:43:05:00:04", "label" : "TT" }
{ "router" : "fe:f0:00:00:04:01", "gateway" : "fe:f1:00:00:04:01", "label" : "TT" }
{ "primary" : "fe:f0:00:00:02:01" }
{ "router" : "fe:f0:00:00:02:01", "neighbor" : "fe:f0:00:00:03:01", "label" : "1.000" }
{ "router" : "fe:f0:00:00:02:01", "neighbor" : "fe:f0:00:00:01:01", "label" : "1.016" }
{ "router" : "fe:f0:00:00:02:01", "neighbor" : "fe:f0:00:00:08:01", "label" : "1.000" }
{ "router" : "fe:f0:00:00:02:01", "gateway" : "fe:f1:00:00:02:01", "label" : "TT" }
{ "router" : "fe:f0:00:00:02:01", "gateway" : "00:00:43:05:00:02", "label" : "TT" }
{ "primary" : "fe:f0:00:00:08:01" }
[...]

License
-------

alfred and vis are licensed under the terms of version 2 of the GNU General
Public License (GPL). Please see the LICENSE file.

The file "packet.h" is an exception and not licensed with the GPL. Instead,
it is licensed using ISC license (see the head of this file). This allows
programs to include this header file (e.g. for communicating with alfred via
unix sockets) without enforcing the restrions of the GPL license on this third
party program.

Contact
-------

As alfred was developed to help on batman-adv, we share communication channels.
Please send us comments, experiences, questions, anything :)

IRC:            #batman   on   irc.freenode.org
Mailing-list:   b.a.t.m.a.n@open-mesh.org (optional  subscription
                at https://lists.open-mesh.org/mm/listinfo/b.a.t.m.a.n)

If you have test reports/patches/ideas, please read the wiki for further instruction
on how to contribute:

http://www.open-mesh.org/projects/open-mesh/wiki/Contribute

You can also contact the Authors:

Simon Wunderlich <siwu@hrz.tu-chmnitz.de>
Sven Eckelmann <sven@open-mesh.com>



debian/changelog
alfred (2013.4.0-1) unstable; urgency=low

  * Initial release

 -- Thomas Hollstegge <thomas.hollstegge@gmail.com>  Thu, 17 Jul 2014 19:47:18 +0000


client.c
/*
 * Copyright (C) 2012 B.A.T.M.A.N. contributors:
 *
 * Simon Wunderlich
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public
 * License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <arpa/inet.h>
#include <sys/select.h>
#include <sys/time.h>
#include <ctype.h>
#include <jansson.h>
#include "alfred.h"

int alfred_client_request_data(struct globals *globals)
{
	unsigned char buf[MAX_PAYLOAD], *pos;
	struct alfred_request_v0 *request;
	struct alfred_push_data_v0 *push;
	struct alfred_status_v0 *status;
	struct alfred_tlv *tlv;
	struct alfred_data *data;
	int ret, len, data_len;
	json_t *root, *value;

	if (unix_sock_open_client(globals, ALFRED_SOCK_PATH))
		return -1;

	request = (struct alfred_request_v0 *)buf;
	len = sizeof(*request);

	request->header.type = ALFRED_REQUEST;
	request->header.version = ALFRED_VERSION;
	request->header.length = sizeof(*request) - sizeof(request->header);
	request->header.length = htons(request->header.length);
	request->requested_type = globals->clientmode_arg;
	request->tx_id = get_random_id();

	ret = write(globals->unix_sock, buf, len);
	if (ret != len)
		fprintf(stderr, "%s: only wrote %d of %d bytes: %s\n",
			__func__, ret, len, strerror(errno));

	root = json_object();

	push = (struct alfred_push_data_v0 *)buf;
	tlv = (struct alfred_tlv *)buf;
	while ((ret = read(globals->unix_sock, buf, sizeof(*tlv))) > 0) {
		if (ret < (int)sizeof(*tlv))
			break;

		if (tlv->type == ALFRED_STATUS_ERROR)
			goto recv_err;

		if (tlv->type != ALFRED_PUSH_DATA)
			break;

		/* read the rest of the header */
		ret = read(globals->unix_sock, buf + sizeof(*tlv),
			   sizeof(*push) - sizeof(*tlv));

		/* too short */
		if (ret < (int)(sizeof(*push) - (int)sizeof(*tlv)))
			break;

		/* read the rest of the header */
		ret = read(globals->unix_sock, buf + sizeof(*push),
			   sizeof(*data));

		data = push->data;
		data_len = ntohs(data->header.length);

		/* would it fit? it should! */
		if (data_len > (int)(sizeof(buf) - sizeof(*push)))
			break;

		/* read the data */
		ret = read(globals->unix_sock,
			   buf + sizeof(*push) + sizeof(*data), data_len);

		/* again too short */
		if (ret < data_len)
			break;

		pos = data->data;

		char id[19];
		sprintf(id, "%02x:%02x:%02x:%02x:%02x:%02x",
		       data->source[0], data->source[1],
		       data->source[2], data->source[3],
		       data->source[4], data->source[5]);

		value = NULL;

		/* Try parsing data as JSON. */
		json_error_t error;
		value = json_loadb((char*)pos, data_len, 0, &error);

		if (value == NULL && memchr(pos, 0, data_len) == NULL) {
			/* data was not JSON and there is no nullbyte in data.
			 * Try expressing data as JSON string. */

			char *tmp;
			tmp = calloc(data_len, sizeof(char));
			strncpy(tmp, (char*)pos, data_len);
			value = json_string(tmp);
			free(tmp);
		}

		if (value == NULL) {
			/* Data could not be expressed as string.
			 * Create an array of integers. */

			value = json_array();
			while(data_len--)
				json_array_append_new(value, json_integer(*pos++));
		}

		json_object_set_new(root, id, value);
	}

	json_dumpf(root, stdout, JSON_INDENT(2));

	json_decref(root);

	unix_sock_close(globals);

	return 0;

recv_err:
	json_decref(root);

	/* read the rest of the status message */
	ret = read(globals->unix_sock, buf + sizeof(*tlv),
		   sizeof(*status) - sizeof(*tlv));

	/* too short */
	if (ret < (int)(sizeof(*status) - sizeof(*tlv)))
		return -1;

	status = (struct alfred_status_v0 *)buf;
	fprintf(stderr, "Request failed with %d\n", status->tx.seqno);

	return status->tx.seqno;;
}

int alfred_client_set_data(struct globals *globals)
{
	unsigned char buf[MAX_PAYLOAD];
	struct alfred_push_data_v0 *push;
	struct alfred_data *data;
	int ret, len;

	if (unix_sock_open_client(globals, ALFRED_SOCK_PATH))
		return -1;

	push = (struct alfred_push_data_v0 *)buf;
	data = push->data;
	len = sizeof(*push) + sizeof(*data);
	while (!feof(stdin)) {
		ret = fread(&buf[len], 1, sizeof(buf) - len, stdin);
		len += ret;

		if (sizeof(buf) == len)
			break;
	}

	push->header.type = ALFRED_PUSH_DATA;
	push->header.version = ALFRED_VERSION;
	push->header.length = htons(len - sizeof(push->header));
	push->tx.id = get_random_id();
	push->tx.seqno = htons(0);

	/* we leave data->source "empty" */
	memset(data->source, 0, sizeof(data->source));
	data->header.type = globals->clientmode_arg;
	data->header.version = globals->clientmode_version;
	data->header.length = htons(len - sizeof(*push) - sizeof(*data));

	ret = write(globals->unix_sock, buf, len);
	if (ret != len)
		fprintf(stderr, "%s: only wrote %d of %d bytes: %s\n",
			__func__, ret, len, strerror(errno));

	unix_sock_close(globals);
	return 0;
}


debian/init.d
#!/bin/sh
### BEGIN INIT INFO
# Provides:          alfred
# Required-Start:    $local_fs $network $remote_fs $syslog
# Required-Stop:     $local_fs $network $remote_fs $syslog
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
### END INIT INFO

# Author: Thomas Hollstegge <thomas.hollstegge@gmail.com>

PATH=/sbin:/usr/sbin:/bin:/usr/bin
DESC="alfred"
NAME=alfred
DAEMON=/usr/sbin/alfred
PIDFILE=/var/run/$NAME.pid
SCRIPTNAME=/etc/init.d/$NAME

# Exit if the package is not installed
[ -x "$DAEMON" ] || exit 0

# Read configuration variable file if it is present
[ -r /etc/default/$NAME ] && . /etc/default/$NAME

# Load the VERBOSE setting and other rcS variables
. /lib/init/vars.sh

# Define LSB log_* functions.
# Depend on lsb-base (>= 3.2-14) to ensure that this file is present
# and status_of_proc is working.
. /lib/lsb/init-functions

#
# Function that starts the daemon/service
#
do_start()
{
	start-stop-daemon --start --quiet --pidfile $PIDFILE --make-pidfile --background --exec $DAEMON --test > /dev/null \
		|| return 1
	start-stop-daemon --start --quiet --pidfile $PIDFILE --make-pidfile --background --exec $DAEMON -- \
		$DAEMON_OPTS \
		|| return 2
}

#
# Function that stops the daemon/service
#
do_stop()
{
	# Return
	#   0 if daemon has been stopped
	#   1 if daemon was already stopped
	#   2 if daemon could not be stopped
	#   other if a failure occurred
	start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --pidfile $PIDFILE --name $NAME
	RETVAL="$?"
	[ "$RETVAL" = 2 ] && return 2
	# Wait for children to finish too if this is a daemon that forks
	# and if the daemon is only ever run from this initscript.
	# If the above conditions are not satisfied then add some other code
	# that waits for the process to drop all resources that could be
	# needed by services started subsequently.  A last resort is to
	# sleep for some time.
	start-stop-daemon --stop --quiet --oknodo --retry=0/30/KILL/5 --exec $DAEMON
	[ "$?" = 2 ] && return 2
	# Many daemons don't delete their pidfiles when they exit.
	rm -f $PIDFILE
	return "$RETVAL"
}

case "$1" in
  start)
	[ "$VERBOSE" != no ] && log_daemon_msg "Starting $DESC" "$NAME"
	do_start
	case "$?" in
		0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
		2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
	esac
	;;
  stop)
	[ "$VERBOSE" != no ] && log_daemon_msg "Stopping $DESC" "$NAME"
	do_stop
	case "$?" in
		0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
		2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
	esac
	;;
  status)
	status_of_proc "$DAEMON" "$NAME" && exit 0 || exit $?
	;;
  restart|force-reload)
	log_daemon_msg "Restarting $DESC" "$NAME"
	do_stop
	case "$?" in
	  0|1)
		do_start
		case "$?" in
			0) log_end_msg 0 ;;
			1) log_end_msg 1 ;; # Old process is still running
			*) log_end_msg 1 ;; # Failed to start
		esac
		;;
	  *)
		# Failed to stop
		log_end_msg 1
		;;
	esac
	;;
  *)
	echo "Usage: $SCRIPTNAME {start|stop|status|restart|force-reload}" >&2
	exit 3
	;;
esac

:


main.c
/*
 * Copyright (C) 2012 B.A.T.M.A.N. contributors:
 *
 * Simon Wunderlich
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public
 * License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <arpa/inet.h>
#include <sys/select.h>
#include <sys/time.h>
#include <signal.h>
#include "alfred.h"

static void alfred_usage(void)
{
	printf("Usage: alfred [options]\n");
	printf("client mode options:\n");
	printf("  -s, --set-data [data type]  sets new data to distribute from stdin\n");
	printf("                              for the supplied data type (0-255)\n");
	printf("  -r, --request [data type]   collect data from the network and prints\n");
	printf("                              it on the network\n");
	printf("  -V, --req-version           specify the data version set for -s\n");
	printf("\n");
	printf("server mode options:\n");
	printf("  -i, --interface             specify the interface to listen on\n");
	printf("  -b                          specify the batman-adv interface configured on the system (default: bat0)\n");
	printf("                              use 'none' to disable the batman-adv based best server selection\n");
	printf("  -m, --master                start up the daemon in master mode, which\n");
	printf("                              accepts data from slaves and synces it with\n");
	printf("                              other masters\n");
	printf("\n");
	printf("  -v, --version               print the version\n");
	printf("  -h, --help                  this help\n");
	printf("\n");
}

static struct globals *alfred_init(int argc, char *argv[])
{
	int opt, opt_ind, i;
	struct globals *globals;
	struct option long_options[] = {
		{"set-data",	required_argument,	NULL,	's'},
		{"request",	required_argument,	NULL,	'r'},
		{"interface",	required_argument,	NULL,	'i'},
		{"master",	no_argument,		NULL,	'm'},
		{"help",	no_argument,		NULL,	'h'},
		{"req-version", required_argument,	NULL,	'V'},
		{"version",	no_argument,		NULL,	'v'},
		{NULL,		0,			NULL,	0},
	};

	globals = calloc(1, sizeof(*globals));
	if (!globals)
		return NULL;

	globals->opmode = OPMODE_SLAVE;
	globals->clientmode = CLIENT_NONE;

	time_random_seed();

	while ((opt = getopt_long(argc, argv, "ms:r:hi:b:vV:", long_options,
				  &opt_ind)) != -1) {
		switch (opt) {
		case 'r':
			globals->clientmode = CLIENT_REQUEST_DATA;
			i = atoi(optarg);
			if (i < ALFRED_MAX_RESERVED_TYPE || i > 255) {
				fprintf(stderr, "bad data type argument\n");
				return NULL;
			}
			globals->clientmode_arg = i;

			break;
		case 's':
			globals->clientmode = CLIENT_SET_DATA;
			i = atoi(optarg);
			if (i < ALFRED_MAX_RESERVED_TYPE || i > 255) {
				fprintf(stderr, "bad data type argument\n");
				return NULL;
			}
			globals->clientmode_arg = i;
			break;
		case 'm':
			globals->opmode = OPMODE_MASTER;
			break;
		case 'i':
			globals->interface = strdup(optarg);
			break;
		case 'b':
			globals->mesh_iface = strdup(optarg);
			break;
		case 'V':
			i = atoi(optarg);
			if (i < 0 || i > 255) {
				fprintf(stderr, "bad data version argument\n");
				return NULL;
			}
			globals->clientmode_version = atoi(optarg);
			break;
		case 'v':
			printf("%s %s\n", argv[0], SOURCE_VERSION);
			printf("A.L.F.R.E.D. - Almighty Lightweight Remote Fact Exchange Daemon\n");
			return NULL;
		case 'h':
		default:
			alfred_usage();
			return NULL;
		}
	}

	if (globals->mesh_iface == NULL)
		globals->mesh_iface = strdup("bat0");

	if (signal(SIGPIPE, SIG_IGN) == SIG_ERR)
		fprintf(stderr, "could not register SIGPIPE handler\n");
	return globals;
}

int main(int argc, char *argv[])
{
	struct globals *globals;
	int ret;

	globals = alfred_init(argc, argv);

	if (!globals)
		return 1;

	switch (globals->clientmode) {
	case CLIENT_NONE:
		ret = alfred_server(globals);
		break;
	case CLIENT_REQUEST_DATA:
		ret = alfred_client_request_data(globals);
		break;
	case CLIENT_SET_DATA:
		ret = alfred_client_set_data(globals);
		break;
	default:
		ret = 0;
	}

	free(globals->interface);
	free(globals->mesh_iface);
	free(globals);

	return ret;
}
